### PR TITLE
feat(pwa) use offline-plugin for service worker setup and offline sup…

### PR DESCRIPTION
**Pull request by EugeneHlushko**
- add offline plugin for utilizing SW and offline support
- takes one off #1400
- the whole website works offline thanks to our existing flatten content tree

Note:
- moved ssg config out as webpack was compiling both in a race (at least on my machine it seemed so.) to now run ssg and then run prod and include ssg run's css that was printed into html files.

### CODE REVIEW

File: webpack.prod.js [hunk 1]
```diff
@@ -43,7 +44,8 @@ const prod = {
         to: './assets/'
       },
       'CNAME'
-    ])
+    ]),
+    new OfflinePlugin()
```
Comment (EugeneHlushko) [inline webpack.prod.js]: Might be worth including a few externals e.g. google fonts?

**Comment by montogeek** [comment]
How does this work?

**Comment by EugeneHlushko** [comment]
It registers a service worker who then catches network requests to assets and caches them. There are defaults enlisted in offline-plugin's readme on options section.

**Comment by EugeneHlushko** [comment]
A good ref: https://github.com/NekR/offline-plugin#docs

**Comment by EugeneHlushko** [comment]
So looks like SSG prod version doesnt need it at all, instead of doing it twice we are doing it once now

**Comment by EugeneHlushko** [comment]
On the preview its still not perfect, for some reason not all css are there when offline

**Comment by montogeek** [comment]
> On the preview its still not perfect, for some reason not all css are there when offline

Good luck in your adventures :D

Services Workers are hard.

**Comment by EugeneHlushko** [comment]
@montogeek ready for review, seem to work as expected

File: package.json [hunk 2]
```diff
@@ -35,7 +35,7 @@
     "fetch:supporters": "node src/utilities/fetch-supporters.js",
     "fetch:starter-kits": "node src/utilities/fetch-starter-kits.js",
     "prebuild": "npm run clean",
-    "build": "run-s fetch content && cross-env NODE_ENV=production webpack --config webpack.prod.js",
+    "build": "run-s fetch content && cross-env NODE_ENV=production webpack --config webpack.ssg.js && cross-env NODE_ENV=production webpack --config webpack.prod.js",
```
Comment (montogeek) [inline package.json]: This `run-s` doesn't work if you append the ENV variable at the beginning?

File: src/components/Site/Site.jsx [hunk 3]
```diff
@@ -77,6 +82,7 @@ class Site extends React.Component {
                   <Route path="/vote" component={Vote} />
                   <Route path="/organization" component={Organization} />
                   <Route path="/starter-kits" component={StarterKits} />
+                  <Route path="/app-shell" component={() => <React.Fragment />} />
```
Comment (montogeek) [inline src/components/Site/Site.jsx]: Why is this needed?

**Comment by montogeek** [comment]
Thanks!

File: src/components/Site/Site.jsx (hunk 3 above)
Comment (EugeneHlushko) [reply src/components/Site/Site.jsx]: This is "empty" state of the page for service worker to serve when offline, our bundle can then render content onto it on clientside. We have two different page templates: home and content pages. Homepage is always saved in SW and this one is used as fall back for the rest to not download all of them to disk when installing service worker, as client doesnt really need SSR version of the page.
